gh-90810: Use code.co_qualname to provide richer information

P403n1x87 opened the pull request:
Use the code.co_qualname field to provide richer information in output generation, such as tracebacks, profiles, debugging etc...

<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:

```
bpo-NNNN: Summary of the changes made
```

Where: bpo-NNNN refers to the issue number in the https://bugs.python.org.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `main`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNN)
```

Where: [X.Y] is the branch name, e.g. [3.6].

GH-NNNN refers to the PR number from `main`.

-->

<!-- gh-issue-number: gh-90810 -->
* Issue: gh-90810
<!-- /gh-issue-number -->

terryjreedy commented on Lib/test/test_logging.py:
```diff
@@ -5049,21 +5049,21 @@ def outer():
         records = self.recording.records
         outer()
-        self.assertEqual(records[-1].funcName, 'innermost')
+        self.assertEqual(records[-1].funcName, 'LoggerTest.test_find_caller_with_stacklevel.<locals>.innermost')
```
This makes the line 112 chars.  Suggest either
```suggestion
        context = 'LoggerTest.test_find_caller_with_stacklevel.<locals>.'
        self.assertEqual(records[-1].funcName, context + 'innermost')
```
or
```suggestion
        self.assertEqual(
            records[-1].funcName,
             'LoggerTest.test_find_caller_with_stacklevel.<locals>.innermost')
```
and repeat below

P403n1x87 replied:
Logging is no longer affected after I restricted the scope of this PR to just tracing. I will address this in a follow up.

terryjreedy commented on Lib/idlelib/stackviewer.py:
```diff
@@ -71,7 +71,7 @@ def GetText(self):
         code = frame.f_code
         filename = code.co_filename
-        funcname = code.co_name
+        funcname = code.co_qualname
```
As IDLE maintainer, I cannot be sure about this without trying some examples.  Even then, other changes are needed in the output and I might prefer to make those first.  In any case, a would want to backport, so a separate PR would be easier.  I have recorded this change as a possibility is my notes on this modue.  I suggest omitting this for now.

P403n1x87 replied:
Thanks for the feedback. I have removed the changes to IDLE and kept only tracing in this PR.

terryjreedy reviewed (changes requested):
See issue for general comments.

One way to split the PR would be separate PRs for changes related to tracebacks, profiles, and 'other'.

P403n1x87 committed:
bpo-46652: Use code.co_qualname to provide richer tracing

Use the code.co_qualname field to provide richer information in tracing
output generation.

pablogsal commented on Lib/test/test_traceback.py:
```diff
@@ -2473,41 +2473,41 @@ def test_exception_group_format(self):
 
         expected = [
                     f'  + Exception Group Traceback (most recent call last):',
-                    f'  |   File "{__file__}", line {lno_g+23}, in _get_exception_group',
```
This is using constantly `self._get_exception_group.__qualname_`, you can capture this into a variable and just use the variable for clarity in the f-strings.

pablogsal replied:
The same apply to other test files.

P403n1x87 replied:
I think this was the only source that could do with a local variable in multiple places?

pablogsal commented on Python/traceback.c:
```diff
@@ -1189,9 +1189,11 @@ dump_frame(int fd, InterpreterFrame *frame)
-       && PyUnicode_Check(code->co_name)) {
-        _Py_DumpASCII(fd, code->co_name);
+    PyObject *name = code->co_qualname;
+    if (name == NULL)
```
PEP 7: Please, add braces.

pablogsal replied:
Alternatively, you can do:

```
PyObject *name = code->co_qualname ? code->co_qualname : code->co_name;
```

pablogsal commented on Python/traceback.c:
```diff
@@ -1189,9 +1189,9 @@ dump_frame(int fd, InterpreterFrame *frame)
-        _Py_DumpASCII(fd, code->co_name);
+    PyObject *name = code->co_qualname ? code->co_qualname : code->co_name;
+    if (name != NULL && PyUnicode_Check(code->co_qualname)) {
+        _Py_DumpASCII(fd, code->co_qualname);
```
This looks wrong, you probably want to do:
```suggestion
    PyObject *name = code->co_qualname ? code->co_qualname : code->co_name;
    if (name != NULL && PyUnicode_Check(name)) {
        _Py_DumpASCII(fd, name);
```

P403n1x87 replied:
Yes! Good catch, thanks!

P403n1x87 committed:
use correct variable in check and branch
